src/Npgsql/NpgsqlRecordDataReader.cs

```diff
@@ -0,0 +1,327 @@
+            }
+
+            var position = Buffer.ReadPosition;
+            return NullableHandler<T>.Exists
+                ? NullableHandler<T>.Read(field.Handler, Buffer, field.Length, fieldDescription: null)
+                : typeof(T) == typeof(object)
+                    ? (T)field.Handler.ReadAsObject(Buffer, field.Length, fieldDescription: null)
+                    : field.Handler.Read<T>(Buffer, field.Length, fieldDescription: null);
+        }
+
+        public override Type GetProviderSpecificFieldType(int ordinal)
+        {
+            var column = CheckRowAndColumn(ordinal);
+            return TypeMapper.GetByOID(column.Oid).GetProviderSpecificFieldType();
+        }
+
+        public override object GetProviderSpecificValue(int ordinal)
+        {
+            var column = CheckRowAndColumnAndSeek(ordinal);
+            if (column.Length == -1)
+                return DBNull.Value;
+            return column.Handler.ReadPsvAsObject(Buffer, column.Length);
+        }
+
+        public override int GetProviderSpecificValues(object[] values)
+        {
+            if (values == null)
+                throw new ArgumentNullException(nameof(values));
+            CheckResultSet();
+
+            var count = Math.Min(FieldCount, values.Length);
+            for (var i = 0; i < count; i++)
+                values[i] = GetProviderSpecificValue(i);
+            return count;
+        }
+
+        public override bool Read()
+        {
+            CheckResultSet();
+
+            Buffer.ReadPosition = _nextRowBufferPos;
+            if (_nextRowIndex >= _numRows)
+                return false;
+
+            if (_nextRowIndex++ != 0)
+                Buffer.ReadInt32(); // Length of record
+
+            var numColumns = Buffer.ReadInt32();
+
+            if (_columns == null || _columns.Length != numColumns)
+                _columns = new ColumnInfo[numColumns];
+
+            for (var i = 0; i < numColumns; i++)
+            {
+                _columns[i].Oid = Buffer.ReadUInt32();
+                _columns[i].BufferPos = Buffer.ReadPosition;
+                var columnLen = Buffer.ReadInt32();
+                if (columnLen >= 0)
+                    Buffer.Skip(columnLen);
+            }
+
+            _nextRowBufferPos = Buffer.ReadPosition;
+
+            return true;
+        }
+
+        ColumnInfo CheckRowAndColumn(int column)
+        {
+            CheckResultSet();
+
+            if (column < 0 || column >= _columns!.Length)
+                throw new IndexOutOfRangeException($"Column must be between {0} and {_columns!.Length - 1}");
+
+            return _columns![column];
+        }
+
+        (NpgsqlTypeHandler Handler, int Length) CheckRowAndColumnAndSeek(int column)
+        {
+            var columnInfo = CheckRowAndColumn(column);
+            Buffer.ReadPosition = columnInfo.BufferPos;
+            var len = Buffer.ReadInt32();
+            return (TypeMapper.GetByOID(columnInfo.Oid), len);
```

Comment:
This does a dictionary lookup for every single field of every row. We can optimize for the case where rows have the same shape, caching the type handler and only looking it up if the type OID actually changed.